lxc/completion.go

```diff
@@ -250,44 +245,144 @@ func (g *cmdGlobal) cmpInstanceKeys(instanceName string) ([]string, cobra.ShellC
+				}
+			}
+		}
+	}
+
+	return configKeys, cmpDirectives | cobra.ShellCompDirectiveNoSpace
 }
 
 // cmpInstanceAllKeys provides shell completion for all possible instance configuration keys.
 // It returns a list of all possible instance configuration keys along with a shell completion directive.
-func (g *cmdGlobal) cmpInstanceAllKeys() ([]string, cobra.ShellCompDirective) {
-	keys := make([]string, 0, len(instancetype.InstanceConfigKeysContainer)+len(instancetype.InstanceConfigKeysVM)+len(instancetype.InstanceConfigKeysAny))
+func (g *cmdGlobal) cmpInstanceAllKeys(profileName string) ([]string, cobra.ShellCompDirective) {
 	cmpDirectives := cobra.ShellCompDirectiveNoFileComp
 
-	for k := range instancetype.InstanceConfigKeysContainer {
-		keys = append(keys, k)
+	// Parse remote
+	resources, err := g.ParseServers(profileName)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	for k := range instancetype.InstanceConfigKeysVM {
-		keys = append(keys, k)
+	resource := resources[0]
+	client := resource.server
+
+	metadataConfiguration, err := client.GetMetadataConfiguration()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	for k := range instancetype.InstanceConfigKeysAny {
-		keys = append(keys, k)
+	instanceConfig, ok := metadataConfiguration.Configs["instance"]
+	if !ok {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	return keys, cmpDirectives
+	// Pre-allocate configKeys slice capacity.
+	keyCount := 0
+	for _, field := range instanceConfig {
+		keyCount += len(field.Keys)
+	}
+
+	configKeys := make([]string, 0, keyCount)
+
+	for _, field := range instanceConfig {
+		for _, key := range field.Keys {
+			for configKey := range key {
+				if strings.HasSuffix(configKey, "*") {
+					configKey = strings.Trim(configKey, "*")
+				}
+
+				configKeys = append(configKeys, configKey)
+			}
+		}
+	}
+
+	return configKeys, cmpDirectives | cobra.ShellCompDirectiveNoSpace
+}
+
+// cmpInstanceSetKeys provides shell completion for instance configuration keys which are currently set.
+// It takes an instance name to determine instance type and returns a list of instance configuration keys along with a shell completion directive.
+func (g *cmdGlobal) cmpInstanceSetKeys(instanceName string) ([]string, cobra.ShellCompDirective) {
+	cmpDirectives := cobra.ShellCompDirectiveNoFileComp
+
+	// Early return when completing server keys.
+	_, instanceNameOnly, found := strings.Cut(instanceName, ":")
+	if instanceNameOnly == "" && found {
+		return g.cmpServerAllKeys(instanceName)
+	}
+
+	resources, err := g.ParseServers(instanceName)
+	if err != nil || len(resources) == 0 {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	resource := resources[0]
+	client := resource.server
+
+	instance, _, err := client.GetInstance(instanceName)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	// Pre-allocate configKeys slice capacity.
+	keyCount := 0
+	keyCount += len(instance.Config)
```

Comment:
```suggestion
	keyCount := len(instance.Config)
```

Reply:
I missed this 😆

Reply:
This came up while copying code across completion functions, which got me thinking that there's a lot of room for refactoring and de-duplicating logic in `lxc/completions.go`. As `lxc/completions.go` has grown, so has duplication. I don't see any immediate benefit to making such improvements, but I'll open an issue and self-assign.